Feature logger

smellilac opened the pull request.

smellilac committed:
Add structured logging with Serilog

smellilac commented:
Git asking to add to name of "LogMessage" method, because because it sees Task in the method return parameter, but I use 
Task.CompletedTask; so it doesn't make sense

HEJOK254 commented:
> Git asking to add to name of "LogMessage" method, because because it sees Task in the method return parameter, but I use 
> Task.CompletedTask; so it doesn't make sense

There seems to be a lot of debate in the c# community regarding the naming of async functions. I chose to go with Microsoft's guidelines, and add the Async suffix to the method names, as DeepSource tells me to do.

The internet has different opinions about adding the Async suffix to Tasks without the async modifier aswell. I'd rename it to LogMessageAsync, but I'm not sure how others like it. I'm not really sure what to do here to be honest

smellilac commented:
> > Git asking to add to name of "LogMessage" method, because because it sees Task in the method return parameter, but I use
> > Task.CompletedTask; so it doesn't make sense
> 
> There seems to be a lot of debate in the c# community regarding the naming of async functions. I chose to go with Microsoft's guidelines, and add the Async suffix to the method names, as DeepSource tells me to do.
> 
> The internet has different opinions about adding the Async suffix to Tasks without the async modifier aswell. I'd rename it to LogMessageAsync, but I'm not sure how others like it. I'm not really sure what to do here to be honest

In this case it is not necessary, since the method does not perform asynchronous operations.
I think it would be more misleading than reflective of the asynchronicity.

HEJOK254 commented:
> In this case it is not necessary, since the method does not perform asynchronous operations.
> I think it would be more misleading than reflective of the asynchronicity.

In this case, the issue can be ignored by adding a comment in the same line, or the line above: `skipcq: CS-R1073`

smellilac committed:
Ignore DeepSource issue (skipcq: CS-R1073)
Fix DeepSource flagging `AutoLog.LogMessage()` as async (CS-R1073)

smellilac commented:
> > In this case it is not necessary, since the method does not perform asynchronous operations.
> > I think it would be more misleading than reflective of the asynchronicity.
> 
> In this case, the issue can be ignored by adding a comment in the same line, or the line above: `skipcq: CS-R1073`

It was good idea :)

HEJOK254 commented:
Sorry for the delay, I'm going to try to review the pr today and suggest some changes if needed

HEJOK254 commented on Program.cs:
```diff
@@ -16,15 +20,21 @@ class Program
+
+        Log.Information("Starting the application.");
+
+        ShowStartMessage();
```
ShowStartMessage() should be enough for the user to know that the program started, so the information log could be changed to a debug log.

The current approach is fine too, just looks a bit weird with a log message before the startup message

HEJOK254 commented on Commands/InteractionServiceHandler.cs:
```diff
@@ -25,7 +26,7 @@ public static async Task InitAsync()
 		}
 		catch
```
Provide additional information about the error
```suggestion
		catch (Exception e)
```

HEJOK254 commented on Commands/InteractionServiceHandler.cs:
```diff
@@ -25,7 +26,7 @@ public static async Task InitAsync()
 		catch
 		{
-			await Program.LogAsync("InteractionServiceHandler", "Error initializing InteractionService", LogSeverity.Critical);
+			Log.Fatal("Error initializing InteractionService");
```
Provide additional information about the error

```suggestion
			Log.Fatal($"Error initializing InteractionService: {e.Message}");
```

HEJOK254 commented on Commands/InteractionServiceHandler.cs:
```diff
@@ -38,7 +39,7 @@ public static async Task RegisterModulesAsync()
 		if (_interactionService == null)
 		{
-			await Program.LogAsync("InteractionServiceManager.RegisterModulesAsync()", "InteractionService not initialized yet", LogSeverity.Error);
+			Log.Error("InteractionService not initialized yet, InteractionServiceManager.RegisterModulesAsync()");
```
I think providing details about where the error comes from is a bit unnecessary, I'm not sure why I did that really lol
What do you think?

```suggestion
			Log.Error("Failed to register modules: InteractionService not initialized.");
```

smellilac replied:
Yes, I also think it is unnecessary.

HEJOK254 commented on Commands/InteractionServiceHandler.cs:
```diff
@@ -52,11 +53,11 @@ public static async Task RegisterModulesAsync()
 			await _interactionService.RegisterCommandsGloballyAsync();
 			_client!.InteractionCreated += OnInteractionCreatedAsync;
-			await Program.LogAsync("InteractionServiceManager", "Modules registered successfully", LogSeverity.Info);
+			Log.Information("Modules registered successfully, InteractionServiceManager");
```
Same situation as above
It could also be simplified to just "Registered modules", but I'm not sure if the "Modules registered successfully" isn't better, as it's clearly indicating a success

```suggestion
			Log.Information("Registered modules");
```
or
```suggestion
			Log.Information("Modules registered successfully");
```

HEJOK254 commented on Commands/InteractionServiceHandler.cs:
```diff
@@ -52,11 +53,11 @@ public static async Task RegisterModulesAsync()
 		catch (Exception e)
 		{
-			await Program.LogAsync("InteractionServiceManager", $"Error registering modules. ({e})", LogSeverity.Critical);
+			Log.Fatal($"Error registering modules. ({e})");
```
Not sure if the user needs the whole error.
Could be dependent on the debug bool from the config for example

```suggestion
			Log.Fatal($"Error registering modules: {e.Message}");
```
Or dependent on the debug bool from the config, though the current implementation would require quite a bit of code
```suggestion
						Log.Fatal($"Error registering modules: {(Program.config != null && Program.config.debug ? e : e.Message)}");
```
Or the additional information could be provided with a separate debug log, requiring less code, but the message would be split into 2.
All of this could be implemented into the other error logs

smellilac replied:
The second option is of course voluminous in terms of code, but extremely correct in terms of its functionality, so it is necessary to choose it.

HEJOK254 commented on Commands/InteractionServiceHandler.cs:
```diff
@@ -66,7 +67,7 @@ public static async Task OnInteractionCreatedAsync(SocketInteraction interaction
 		if (_interactionService == null)
 		{
-			await Program.LogAsync("InteractionServiceManager.OnInteractionCreatedAsync()", "InteractionService not initialized yet", LogSeverity.Error);
+			Log.Error("InteractionService not initialized yet, InteractionServiceManager.OnInteractionCreatedAsync()");
```
Same as in `RegisterModulesAsync()`

```suggestion
			Log.Error("Error handling interaction: InteractionService not initialized.");
```

HEJOK254 commented on Commands/InteractionServiceHandler.cs:
```diff
@@ -79,7 +80,7 @@ public static async Task OnInteractionCreatedAsync(SocketInteraction interaction
 		catch (Exception e)
 		{
-			await Program.LogAsync("InteractionServiceManager", $"Error handling interaction. {e.Message}", LogSeverity.Error);
+			Log.Error($"Error handling interaction. {e.Message}");
```
Just a small change for better log message consistency

```suggestion
			Log.Error($"Error handling interaction: {e.Message}");
```

HEJOK254 commented on Commands/Modules/VideoUtils.cs:
```diff
@@ -69,7 +70,7 @@ public class VideoUtils : InteractionModuleBase
 		{
 			Directory.CreateDirectory("./tmp");
-			await Program.LogAsync("VideoUtils", "TMP directory not found. Created it automatically.", LogSeverity.Info);
+			Log.Information("TMP directory not found. Created it automatically, VideoUtils");
```
Same thing as in `InteractionServiceHandler.cs` basically. Especially since this is normally a one time message, and it's only present in one class.

```suggestion
			Log.Information("TMP directory not found. Created it automatically");
```

HEJOK254 commented on ConfigManager.cs:
```diff
@@ -14,8 +15,8 @@ public class Config
 		{
-			Program.LogAsync("Config", $"Config file not found at: {path}", LogSeverity.Critical);
-			return null;
+            Log.Fatal("Config file not found at: {Path}", path);
```
The message doesn't actually return a path, but just says "./config.json", which doesn't help fix the problem
I think I actually forgot to test it in my original code, so I guess it can be fixed now :P

```suggestion
            Log.Fatal($"Config file not found at: {Path.GetFullPath(path)}");
```

HEJOK254 commented on ConfigManager.cs:
```diff
@@ -24,8 +25,8 @@ public class Config
 		}
 		catch
```
Provide more information

```suggestion
		catch (Exception e)
```

HEJOK254 commented on ConfigManager.cs:
```diff
@@ -24,8 +25,8 @@ public class Config
 		{
-			Program.LogAsync("Config", "Failed to parse config file.", LogSeverity.Critical);
-			return null;
+            Log.Fatal("Failed to parse config file.");
```
Use with the change above
Provide more information about the issue, so the user can have a general idea of what the problem is

```suggestion
            Log.Fatal($"Failed to parse config file: {e.Message}");
```

HEJOK254 commented on LoggerConfig/LoggerImplementation/AutoLog.cs:
```diff
@@ -0,0 +1,31 @@
+
+public class AutoLog
+{
+    // skipcq: CS-R1073
```
Explain why the error is skipped

```suggestion
    // LogMessage does not run asynchronously, so we can ignore the DeepSource error
    // skipcq: CS-R1073
```

HEJOK254 commented on LoggerConfig/LoggerImplementation/AutoLog.cs:
```diff
@@ -0,0 +1,31 @@
+
+namespace QuickEdit.LoggerConfig.LoggerImplementation;
+
+public class AutoLog
```
I think `AutoLog` sounds a bit weird, but I honestly can't think of anything else except for like `DiscordLogger`, which would kind of sound like it's logging actual messages from discord, which would sound a bit like the bot is "spying"

The namespace makes `AutoLog` sound better though, so I think it could be left like this

HEJOK254 commented on LoggerConfig/LoggerImplementation/AutoLog.cs:
```diff
@@ -0,0 +1,31 @@
+public class AutoLog
+{
+    // skipcq: CS-R1073
+    public static Task LogMessage(LogMessage message)
```
`LogMessage()` sounds a bit generic but once again, I can't think of any other good name, so I think it's fine

HEJOK254 commented on LoggerConfig/LoggerImplementation/AutoLog.cs:
```diff
@@ -0,0 +1,31 @@
+using System.Text;
+using System.Threading.Tasks;
+
+namespace QuickEdit.LoggerConfig.LoggerImplementation;
```
This class doesn't really have anything to do with configuration, so I think changing the namespace both here and in `SerilogConfiguration.cs` to something like `QuickEdit.Logger` or `QuickEdit.Logging` *could* be better? I'm not sure though

HEJOK254 commented on LoggerConfig/SerilogConfiguration.cs:
```diff
@@ -0,0 +1,29 @@
+
+        var logPath = Path.Combine(intermediateOutputPath, $"consoleapplog-{DateTime.UtcNow:yyyyMMddHHmmss}.txt");
+
+        Log.Logger = new LoggerConfiguration()
+            .MinimumLevel.Information()
+            .Enrich.WithMachineName()
+            .WriteTo.Console()
+            .WriteTo.File(logPath, rollingInterval: RollingInterval.Day)
+            .CreateLogger();
```
The logger should probably log debug messages when debug is on in the configuration

I'm not very familiar with Serilog just yet, so my implementation of this might not be very good, but this is what I came up with:
```suggestion
	var loggerConfig = new LoggerConfiguration()
		.Enrich.WithMachineName()
		.WriteTo.Console()
		.WriteTo.File(logPath, rollingInterval: RollingInterval.Day);

	if (Program.config != null && Program.config.debug)
		loggerConfig.MinimumLevel.Debug();
	else
		loggerConfig.MinimumLevel.Information();
		
	Log.Logger = loggerConfig.CreateLogger();
```

HEJOK254 reviewed (changes requested):
I've reviewed the changes, and everything seems to be working great, however I made come suggestions here and there.

Keep in mind that these are just suggestions :)

smellilac commented:
You don't merge PR.
Want to add/correct something?

HEJOK254 commented:
> You don't merge PR.
> Want to add/correct something?

Sorry, I've been out of town for a few days, so I wasn't able to review it

I should be back in a few days, and I'll try to review it when I can

Sorry for the delay

HEJOK254 commented on QuickEdit.Logger/LoggerImplementation/AutoLog.cs:
I meant the namespace in the file, not the folder name, sorry :/

So the folder would be called Logger (and the file path would be `/Logger/AutoLog.cs`), while the namespace on line 10 could be for example `QuickEdit.Logger`

smellilac replied:
> I meant the namespace in the file, not the folder name, sorry :/
> 
> So the folder would be called Logger (and the file path would be `/Logger/AutoLog.cs`), while the namespace on line 10 could be for example `QuickEdit.Logger`

It's okay, misunderstandings happen to everyone :relaxed:

I'll fix it now

HEJOK254 commented on QuickEdit.Logger/SerilogConfiguration.cs:
I meant the namespace in the file, not the folder name, sorry :/

So the folder would be called Logger (and the file path would be `/Logger/SerilogConfiguration.cs`), while the namespace on line 9 could be for example `QuickEdit.Logger`

HEJOK254 reviewed (changes requested):
I've reviewed the changes, sorry for the delay

smellilac commented:
> delay

No problem. Glad to have you back!

smellilac commented:
I renamed the main folder to "Logger" and deleted the "LoggerImplementation" folder.

I hope we understand each other correctly now, if not, please let me know.

HEJOK254 commented on Logger/SerilogConfiguration.cs:
```diff
@@ -0,0 +1,34 @@
+using System.Text;
+using System.Threading.Tasks;
+
+namespace QuickEdit.LoggerConfig;
```
Use the same namespace as `AutoLog.cs`, as there isn't really a reason to separate the two
```suggestion
namespace QuickEdit.Logger;
```

HEJOK254 commented on Program.cs:
```diff
@@ -3,6 +3,10 @@
 using FFMpegCore.Helpers;
 using QuickEdit.Commands;
+using QuickEdit.Logger;
+using QuickEdit.LoggerConfig;
```
If `AutoLog` is moved to the `QuickEdit.Logger` namespace, the `QuickEdit.LoggerConfig` using can be removed
```suggestion
```

HEJOK254 reviewed (changes requested):
It's all good now, but I think `AutoLog` and `SerilogConfiguration` should be in the same namespace, since they're only 2 classes, and they're related

HEJOK254 reviewed (dismissed):
Looks good!

I'd clean up the commit history a bit by doing an interactive rebase if you're ok with that. It would basically rewrite the commit history, in such a way that, for example, some of the small commits fixing an issue with the previous commit, which are quite common when working on a feature, could be merged into one, cleaning up the history a bit before getting merged.

This would also require a force push, to rewrite the commits. I always make sure to keep a backup of the branch, though, in case something goes wrong.

After that, I would slightly clean up all the files, like for example fixing some broken indentations in some places probably caused by me forgetting to add them in the review suggestions :P

I'm not an expert at git, and this is pretty much my first time collaborating with other people on GitHub (thanks :)), which is why I'm always asking about everything. I can just leave the commit history as is and merge the PR, or squash merge it if needed, but I just thought it would be better to tidy it up a bit.

smellilac commented:
> Looks good!
> 
> I'd clean up the commit history a bit by doing an interactive rebase if you're ok with that. It would basically rewrite the commit history, in such a way that, for example, some of the small commits fixing an issue with the previous commit, which are quite common when working on a feature, could be merged into one, cleaning up the history a bit before getting merged.
> 
> This would also require a force push, to rewrite the commits. I always make sure to keep a backup of the branch, though, in case something goes wrong.
> 
> After that, I would slightly clean up all the files, like for example fixing some broken indentations in some places probably caused by me forgetting to add them in the review suggestions :P
> 
> I'm not an expert at git, and this is pretty much my first time collaborating with other people on GitHub (thanks :)), which is why I'm always asking about everything. I can just leave the commit history as is and merge the PR, or squash merge it if needed, but I just thought it would be better to tidy it up a bit.

Yes, indeed, the PR looks dirty, so I'm all for you cleaning it up.

I'm also far from being a professional in git and this is also my first work in collaboration (thank you very much too!), but as far as I understand, I can't help with cleaning the commits, since I don't have the necessary rights in your repository. However, I can clean the code and upload the commit.

After which you will start cleaning, okay?

HEJOK254 commented:
> However, I can clean the code and upload the commit.

Yeah, you can do that if you want.
I was basically thinking of fixing the indentations and formatting, plus maybe removing the unused `using` directives. All of that can be achieved by running a clean-up in Visual Studio, as far as I'm aware, and additionally double-checking if everything is correct and works properly.

I can do it if you don't want to, and as for the history clean-up, I wanted to do it in this pull request, so on your fork, in your branch, as I think I should have the necessary permissions for it, but I'm not sure yet.

After that, I'll merge the pr

smellilac commented:
> After that, I'll merge the pr

Ok, I didn't know it was so easy to clean up code from extra spaces and unused directives.

In that case, I'm waiting for your code cleaning, as well as cleaning up commits.

HEJOK254 committed:
Code Cleanup

Fixed things like:
 - indentation
 - formatting
 - file format (Central European (Windows) -> UTF-8)
 - access modifiers (program class is now internal)
 - removed unused `using` directives

smellilac committed:
Add option to enable debug logging

Co-authored-by: Jakub Dębski <90698026+HEJOK254@users.noreply.github.com>

smellilac committed:
Improve log messages
Improved log messages by:
 - Adding more error information to the log messages
 - Making them less verbose
 - Adding additional debug information in debug mode (config.json)

Co-authored-by: Jakub Dębski <90698026+HEJOK254@users.noreply.github.com>

smellilac committed:
Remove startup log

smellilac committed:
Change logger namespace

Co-authored-by: Jakub Dębski <90698026+HEJOK254@users.noreply.github.com>

HEJOK254 commented:
I pushed the changes to [HEJOK254/Discord-QuickEdit/feature/logger-preview](https://github.com/HEJOK254/Discord-QuickEdit/tree/feature/logger-preview), instead of force pushing it to your branch, so we can discuss the changes and change things if needed, without having to force push on your branch all the time.

Please check if everything looks ok, or if I should rename/delete/squash etc. some commits. I tried to keep the commits small, which should help in troubleshooting in case one of them introduced some issue, while also grouping together changes regarding the same "feature", like for example improving the log messages.

If you have any suggestions, let me know, and I'll be happy to change it

smellilac commented:
> If you have any suggestions, let me know, and I'll be happy to change it
I looked at the commits in this branch. They look clean and structured. I have no ideas/wishes, in my opinion you did everything well.

smellilac commented:
> > If you have any suggestions, let me know, and I'll be happy to change it

 I looked at the commits in this branch. They look clean and structured. I have no ideas/wishes, in my opinion you did everything well.

HEJOK254 commented on Logger/SerilogConfiguration.cs:
```diff
@@ -0,0 +1,29 @@
+
+	public static void ConfigureLogger()
+	{
+		var intermediateOutputPath =
+			Path.GetFullPath(Path.Combine(AppDomain.CurrentDomain.BaseDirectory, @"..\..\..", @"obj\Debug\net8.0\Logs"));
+
+		Directory.CreateDirectory(intermediateOutputPath);
+
+		var logPath = Path.Combine(intermediateOutputPath, $"consoleapplog-{DateTime.UtcNow:yyyyMMddHHmmss}.txt");
```
The log is saved in a location only appropriate in Visual Studio. On a server, it would create the logs in a hard to find path, and might cause confusion.

I'd recommend changing it to just `logs`, as it should be clear to the user where the logs are located.
That would also eliminate the need for the `intermediateOutputPath` variable, as just "logs" could be used, though a better approach, I think, is still using a variable, but renaming it.

Another thing I've noticed, is the log names contain the date twice, and are not really specifying what program they're from. From my testing, it seems that Serilog (or something else idk) automatically adds the date to the filename, removing the need for manually adding the date and time.

```suggestion
		var logDirectory = "logs";

		Directory.CreateDirectory(logDirectory);

		var logPath = Path.Combine(logDirectory, $"quickedit-.log");
```

HEJOK254 commented on Logger/SerilogConfiguration.cs:
```diff
@@ -0,0 +1,29 @@
+		var logPath = Path.Combine(intermediateOutputPath, $"consoleapplog-{DateTime.UtcNow:yyyyMMddHHmmss}.txt");
+
+		var loggerConfig = new LoggerConfiguration()
+			.Enrich.WithMachineName()
```
I don't think this enrichment is used at all here. To use it, I think a custom `outputTemplate` should be used. However, I don't think this feature is necessary in this case, so it could just be removed along with the `Serilog.Enrichers.Environment` NuGet package.

```suggestion
```

If you think it should stay, let me know

HEJOK254 reviewed (changes requested):
While reviewing the code before merging, a few things have caught my eye, that I think should be fixed

Sorry I'm noticing about it just now

smellilac committed:
Logging fix

- Removed log enrichment
- Changed the path to the log file to the universal directory

Co-authored-by: Jakub Dębski <90698026+HEJOK254@users.noreply.github.com>

HEJOK254 committed:
Remove unused Serilog enricher (NuGet)

Since 76c6a1 the package `Serilog.Enrichers.Environment` is not needed anymore, as it's unused

Forgot to remove it while reviewing the PR whoops

HEJOK254 committed:
Fix DeepSource issue (CS-W1000)

Empty interpolated string (CS-W1000)

Forgot to remove that `$` while reviewing the PR whoops (76c6a12ce767fbd81e7661b24f9698cd336cbde3)

HEJOK254 commented:
I forgot to remove that `$` sign in `SerilogConfiguration.cs` in the review suggestion, which was causing the DeepSource issue, and I also forgot to add a suggestion to remove the Serilog Enricher, which is now unused, so I went ahead and committed the fixes myself, hope that's not an issue

Anyway, I think the PR is ready to be merged. If you have any questions or any more changes you would like to make, let me know, if not, I'll merge it soon :)

smellilac commented:
> Anyway, I think the PR is ready to be merged. If you have any questions or any more changes you would like to make, let me know, if not, I'll merge it soon :)

No questions or suggestions.

Thank you for working together!

HEJOK254 commented:
> No questions or suggestions.

Ok, I'll merge it now

Thanks for helping me with the project :)
